daemon/runtime_unix.go

```diff
@@ -47,92 +63,88 @@ func defaultV2ShimConfig(conf *config.Config, runtimePath string) *types.ShimCon
+}
 
-		if err = os.Rename(runtimeDir, runtimeDir+"-old"); err != nil {
-			return
-		}
-		if err = os.Rename(tmpDir, runtimeDir); err != nil {
-			err = errors.Wrap(err, "failed to setup runtimes dir, new containers may not start")
-			return
-		}
-		if err = os.RemoveAll(runtimeDir + "-old"); err != nil {
-			logrus.WithError(err).WithField("dir", tmpDir).
-				Warn("failed to remove old runtimes dir")
-		}
-	}()
+func setupRuntimes(cfg *config.Config) (runtimes, error) {
+	newrt := runtimes{
+		configured: make(map[string]*shimConfig),
+	}
 
+	dir := runtimeScriptsDir(cfg)
 	for name, rt := range cfg.Runtimes {
-		if len(rt.Args) > 0 {
-			script := filepath.Join(tmpDir, name)
-			content := fmt.Sprintf("#!/bin/sh\n%s %s $@\n", rt.Path, strings.Join(rt.Args, " "))
-			if err := os.WriteFile(script, []byte(content), 0700); err != nil {
-				return err
+		binaryName := rt.Path
+		needsWrapper := len(rt.Args) > 0
+		if needsWrapper {
+			var err error
+			binaryName, err = wrapRuntime(dir, name, rt.Path, rt.Args)
+			if err != nil {
+				return runtimes{}, err
 			}
 		}
-		if rt.Shim == nil {
-			rt.Shim = defaultV2ShimConfig(cfg, rt.Path)
+		c := defaultV2ShimConfig(cfg, binaryName)
+		if needsWrapper {
+			c.PreflightCheck = func() error {
+				// Check that the runtime path actually exists so that we can return a well known error.
+				_, err := exec.LookPath(rt.Path)
+				return errors.Wrap(err, "error while looking up the specified runtime path")
+			}
 		}
-	}
-	return nil
-}
-
-// rewriteRuntimePath is used for runtimes which have custom arguments supplied.
-// This is needed because the containerd API only calls the OCI runtime binary, there is no options for extra arguments.
-// To support this case, the daemon wraps the specified runtime in a script that passes through those arguments.
-func (daemon *Daemon) rewriteRuntimePath(cfg *config.Config, name, p string, args []string) (string, error) {
-	if len(args) == 0 {
-		return p, nil
+		newrt.configured[name] = c
 	}
 
-	// Check that the runtime path actually exists here so that we can return a well known error.
-	if _, err := exec.LookPath(p); err != nil {
-		return "", errors.Wrap(err, "error while looking up the specified runtime path")
-	}
-
-	return filepath.Join(cfg.Root, "runtimes", name), nil
+	return newrt, nil
 }
 
-func (daemon *Daemon) getRuntime(cfg *config.Config, name string) (*types.Runtime, error) {
-	rt := cfg.GetRuntime(name)
-	if rt == nil {
-		if !config.IsPermissibleC8dRuntimeName(name) {
-			return nil, errdefs.InvalidParameter(errors.Errorf("unknown or invalid runtime name: %s", name))
-		}
-		return &types.Runtime{Shim: &types.ShimConfig{Binary: name}}, nil
+// A non-standard Base32 encoding which lacks vowels to avoid accidentally
+// spelling naughty words. Don't use this to encode any data which requires
+// compatibility with anything outside of the currently-running process.
+var base32Disemvoweled = base32.NewEncoding("0123456789BCDFGHJKLMNPQRSTVWXYZ-")
```

Comment:
:joy: :heart:

(Should this also remove "vowel looking" numbers like `0`, `1`, `3`, `4`?)